src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs

```diff
@@ -90,99 +72,147 @@ internal sealed class TupleTypeSymbol : NamedTypeSymbol
+                    typeArgumentsBuilder.Add(new TypeWithModifiers(arguments[i], modifiers[i]));
+                }
+                else
+                {
+                    typeArgumentsBuilder.Add(new TypeWithModifiers(arguments[i]));
+                }
             }
 
-            _fields = fieldsBuilder.ToImmutableAndFree();
+            if (!modifiers.IsDefault && !modifiers[RestPosition - 1].IsDefaultOrEmpty)
+            {
+                typeArgumentsBuilder.Add(new TypeWithModifiers(extensionTuple, modifiers[RestPosition - 1]));
+            }
+            else
+            {
+                typeArgumentsBuilder.Add(new TypeWithModifiers(extensionTuple));
+            }
+
+            tupleCompatibleType = tupleCompatibleType.ConstructedFrom.Construct(typeArgumentsBuilder.ToImmutable(), unbound: false);
+            return tupleCompatibleType;
         }
 
         /// <summary>
         /// Copy this tuple, but modify it to use the new underlying type.
         /// </summary>
         internal TupleTypeSymbol WithUnderlyingType(NamedTypeSymbol newUnderlyingType)
         {
-            Debug.Assert((object)newUnderlyingType.OriginalDefinition == (object)UnderlyingTupleType.OriginalDefinition);
+            Debug.Assert((object)newUnderlyingType.OriginalDefinition == (object)_underlyingType.OriginalDefinition);
 
-            return new TupleTypeSymbol(this, newUnderlyingType);
+            return Create(_locations, newUnderlyingType, _elementLocations, _elementNames);
         }
 
         /// <summary>
         /// Copy this tuple, but modify it to use the new element names.
         /// </summary>
         internal TupleTypeSymbol WithElementNames(ImmutableArray<string> newElementNames)
         {
-            Debug.Assert(newElementNames.IsDefault || this._fields.Length == newElementNames.Length);
+            Debug.Assert(newElementNames.IsDefault || this._elementTypes.Length == newElementNames.Length);
 
-            var originalFields = _fields;
-            for (int i = 0; i < originalFields.Length; i++)
+            if (this._elementNames.IsDefault)
             {
-                var originalField = originalFields[i];
-                var originalName = originalField.Name;
-                var newName = GetFieldNameFromArrayOrDefaultName(newElementNames, i);
-
-                if (originalName != newName)
+                if (newElementNames.IsDefault)
                 {
-                    // at least one name is different
-                    return new TupleTypeSymbol(this, newElementNames);
+                    return this;
                 }
             }
+            else if (!newElementNames.IsDefault && this._elementNames.SequenceEqual(newElementNames))
+            {
+                return this;
+            }
 
-            // all names are the same
-            return this;
+            return new TupleTypeSymbol(null, _underlyingType, default(ImmutableArray<Location>), newElementNames, _elementTypes);
```

Comment:
This is an open issue tracked by the PROTOTYPE comment at the call site, I believe.

---
In reply to: [61810622](https://github.com/dotnet/roslyn/pull/11006#discussion_r61810622) [](ancestors = 61810622)